### PR TITLE
Fix Hive support in Spark 2.3

### PR DESCRIPTION
**Pull request by kevinykuo**
- `spark_connect()` now first initializes a `SparkSession` before obtaining a `SparkContext` from it. This is required in order to properly set the static SQL configs such as `spark.sql.catalogImplementation`. The `SparkSession` is immediately cached as the `hive_context` field.
- Default `spark.sql.catalogImplementation` to `hive` so Hive support is enabled. Users can opt out of this by setting the option to `in-memory`.
- `spark_version()` now actually caches the Spark version in the connection object.

Closes https://github.com/rstudio/sparklyr/issues/1496, closes https://github.com/rstudio/sparklyr/issues/415

### CODE REVIEW

File: R/shell_connection.R [hunk 1]
```diff
@@ -546,34 +546,82 @@ initialize_connection.spark_shell_connection <- function(sc) {
...
+          }
+        )
+
+        ctx
+      }
 
       invoke(backend, "setSparkContext", sc$spark_context)
     }
 
-    sc$spark_context$connection <- sc
```
Comment (javierluraschi) [inline R/shell_connection.R]: We probably still need this one and the `java_context` bellow.

When we call `invoke_static()` with an `sc`, the `sc` becomes attached to the object, this is to be able to execute `invoke()` on the object etc. This is fine for most operations; however, if you need to get the `spark_context()` and then indirectly get the `spark_context()` from there again, you would get a `spark_context()` that is no longer there. The way we patched this was by reattaching the connection once the `spark_context()` is created.

File: R/spark_version.R [hunk 2]
```diff
@@ -18,8 +18,12 @@ spark_version_clean <- function(version) {
 spark_version <- function(sc) {
 
   # use cached value if available
-  if (!is.null(sc$spark_version))
-    return(sc$spark_version)
+  if (!is.null(sc$state$spark_version))
```
Comment (javierluraschi) [inline R/spark_version.R]: This is nice! Get's called a bunch of times! I had to do this for Livy already (see https://github.com/rstudio/sparklyr/blob/3a3142d519026e294dc0f50d400cf052c5f8d1a6/R/livy_connection.R#L778),but it doesn't hurt to do this for shell connections as well! 👍

**Comment by javierluraschi** [comment]
Ah, yes, so regarding the broken test, the second code review comment about keeping the re-assignment of `sc` will probably fix this failure:

```
Unable to retrieve a spark_connection from object of class NULL
1: df1_tbl %>% sdf_broadcast() %>% left_join(df2_tbl, by = "b") %>% spark_dataframe() %>% 
       invoke("queryExecution") %>% invoke("optimizedPlan") %>% invoke("toString") at testthat/test-dplyr.R:118
2: withVisible(eval(quote(`_fseq`(`_lhs`)), env, env))
3: eval(quote(`_fseq`(`_lhs`)), env, env)
4: eval(quote(`_fseq`(`_lhs`)), env, env)
5: `_fseq`(`_lhs`)
6: freduce(value, `_function_list`)
7: function_list[[i]](value)
8: sdf_broadcast(.)
9: invoke_static(sc, "org.apache.spark.sql.functions", "broadcast", sdf) %>% sdf_register()
10: withVisible(eval(quote(`_fseq`(`_lhs`)), env, env))
...
28: `_fseq`(`_lhs`)
29: freduce(value, `_function_list`)
30: function_list[[i]](value)
31: invoke(., "sql", as.character(sqlFields))
32: invoke_trace(spark_connection(jobj), "Invoking", method)
33: spark_config_value(sc$config, "sparklyr.invoke.trace", FALSE)
34: name %in% names(config)
35: spark_connection(jobj)
36: spark_connection.default(jobj)
37: stop("Unable to retrieve a spark_connection from object of class ", paste(class(x), 
       collapse = " "), call. = FALSE)
```

**Comment by javierluraschi** [comment]
The Windows test might be related as well to the previous comment, but not sure, might be something else worth investigating. Here is the output for convenience:

```
-- 1. Error: 'spark_connect' can provide a 'spark_log' (@test-connect-shell.R#15
cannot open the connection
1: capture.output({
       spark_log(sc)
   }) at testthat/test-connect-shell.R:15
2: evalVis(expr)
3: withVisible(eval(expr, pf))
4: eval(expr, pf)
5: eval(expr, pf)
6: spark_log(sc)
7: spark_log.spark_shell_connection(sc)
8: readLines(log)

/usr/bin/tar: Cannot connect to C: resolve failed
/usr/bin/tar: Cannot connect to C: resolve failed
== testthat results  ===========================================================
OK: 1849 SKIPPED: 22 FAILED: 1
1. Error: 'spark_connect' can provide a 'spark_log' (@test-connect-shell.R#15) 
```

File: R/shell_connection.R [hunk 3]
```diff
@@ -546,34 +546,82 @@ initialize_connection.spark_shell_connection <- function(sc) {
...
 
       # create the spark context and assign the connection to it
-      sc$spark_context <- invoke_static(
-        sc,
-        "org.apache.spark.SparkContext",
-        "getOrCreate",
-        conf
-      )
+
+      sc$spark_context <- if (spark_version(sc) >= "2.0") {
```
[the remark this answers is not in the document]
Comment (kevinykuo) [reply R/shell_connection.R]: The "Hive context" for Spark 2.0+ is really the `SparkSession`, which we need anyway, so I don't think we're losing performance here, no?

File: R/shell_connection.R [hunk 4]
```diff
@@ -546,34 +547,82 @@ initialize_connection.spark_shell_connection <- function(sc) {
...
-        sc,
-        "org.apache.spark.SparkContext",
-        "getOrCreate",
-        conf
-      )
 
-      invoke(backend, "setSparkContext", sc$spark_context)
-    }
+      sc$state$spark_context <- if (spark_version(sc) >= "2.0") {
+
```
Comment (kevinykuo) [inline R/shell_connection.R]: We need to create a `SparkSession` anyway so it doesn't hurt performance to cache.

File: R/shell_connection.R [hunk 5]
```diff
@@ -453,26 +453,27 @@ connection_is_open.spark_shell_connection <- function(sc) {
 
 #' @export
 spark_log.spark_shell_connection <- function(sc, n = 100, filter = NULL, ...) {
-  if (!is.null(sc$output_file) && file.exists(sc$output_file)) {
```
Comment (kevinykuo) [inline R/shell_connection.R]: Consolidate `spark_log.spark_shell_connection()` and `spark_log.spark_connection()`.

**Comment by kevinykuo** [comment]
@javierluraschi tests are passing now, could you take another quick look?

File: R/spark_gateway.R [hunk 6]
```diff
@@ -57,7 +57,7 @@ spark_gateway_connection <- function(master, config, gatewayInfo, gatewayAddress
   })
 
   # create the shell connection
-  sc <- structure(class = c("spark_connection", "spark_gateway_connection", "spark_shell_connection"), list(
```
Comment (mengxr) [inline R/spark_gateway.R]: This change broke Databricks connection. Our test against master failed with:

~~~
Error in value[[3L]](cond) : 
  Failed during initialize_connection:Failed during initialize_connection: Unable to retrieve a spark_connection from object of class NULL
~~~

The error is from `spark_connection.default`. With the new refactor, we might need an implementation of `spark_connection` for Databricks.

@falaki @javierluraschi @kevinykuo Any recommendation for the fix?

**Pull request merged**